File: Smart_Grid_System -  simulation  python/smartgrid.py
import csv

class SmartGrid:
    def __init__(self, vertices):
        self.places = vertices
        self.Adjlist = [[] for _ in range(vertices)]
        self.path = [0] * (vertices + 1)
        self.voltpath = [0.0] * (vertices + 1)
        self.temp_distance = 0
        self.rene_res = ['solar Power','wind Energy','Biomass','Hydro Power']
        self.nonrene_res = ['coal','Natural Gas','Oil','Nuclear Power']
        self.placeMap = {"grid": 0}
        self.voltdict = {30: 33.0, 50: 48.0, 60: 50.0, 80: 60.0, 90: 66.0, 20: 0.44, 70: 49.0, 40: 40.0,
                         85: 63.0, 120: 345.0, 100: 345.0, 200: 400.0, 300: 450.0, 400: 490.0, 190: 380.0, 160: 360.0}
        self.place_names = [""] * vertices

    class Container:
        def __init__(self, s, d, dist, volt, from_, to):
            self.source = s
            self.destination = d
            self.distance = dist
            self.voltage = volt
            self.from_ = from_
            self.to = to

    def add_edge(self, src, des, dist, volt, from_, to):
        c = self.Container(src, des, dist, volt, from_, to)
        self.Adjlist[src].append(c)

    def remove_edge(self, src, des):
        for neigh in self.Adjlist[src]:
            if neigh.destination == des:
                self.Adjlist[src].remove(neigh)
                print("edge removed:", src, "---", des)
                break

    def first_find_path(self, src, des, p_src):
        for neighbor in self.Adjlist[src]:
            self.temp_distance += neighbor.distance
            self.path[p_src] = neighbor.destination
            self.voltpath[p_src] = neighbor.voltage
            if neighbor.destination == des:
                return self.temp_distance
            dist = self.first_find_path(neighbor.destination, des, p_src + 1)
            if dist == 0:
                self.temp_distance -= neighbor.distance
                self.path[p_src] = 0
                self.voltpath[p_src] = 0
            else:
                return dist
        return 0

    def second_find_path(self, src, des):
        for neighbor in self.Adjlist[src]:
            self.temp_distance += neighbor.distance
            if neighbor.destination == des:
                return self.temp_distance
            dist = self.second_find_path(neighbor.destination, des)
            if dist == 0:
                self.temp_distance -= neighbor.distance
            else:
                return dist
        return 0

    def grid_optimization(self, s, d, dist):
        for i in range(self.places):
            self.path[i] = 0

        for i in range(self.places):
            self.voltpath[i] = 0.0

        getvolt = self.voltdict[dist]

        min_distance = self.first_find_path(0, d, 1)
        self.temp_distance = 0
        org_dis = self.second_find_path(0, s)
        sum_ = org_dis + dist

        if sum_ > min_distance:
            print("min-distance:", "0--", d, "is", min_distance)
        else:
            print("min-distance:", "0--", d, "is", sum_)
            j = self.places - 1
            while j > 0:
                if self.path[j] == 0:
                    j -= 1
                else:
                    self.remove_edge(self.path[j - 1], self.path[j])
                    self.add_edge(s, d, dist, getvolt, 'c', 'c')
                    print("edge added:", s, "---", d)
                    j = 0

    def print_grid(self):
        print("S  d      dis      volt")
        for i in range(self.places):
            for neigh in self.Adjlist[i]:
                print("(" + str(neigh.source) + ", " + str(neigh.destination) + ")" + " --> " + str(neigh.distance) +
                      " <--> " + str(neigh.voltage))
        for key, value in self.placeMap.items():
            print("Key:", key, ", Value:", value)

    def contains_value(self, key):
        return key in self.placeMap

    def add_place(self, name, num):
        self.placeMap[name] = num

    def store_graph_to_file(self, filename):
        with open(filename, 'w') as file:
            writer = csv.writer(file)
            for i in range(len(self.Adjlist)):
                edges = self.Adjlist[i]
                for edge in edges:
                    writer.writerow([edge.source, edge.destination, edge.distance, edge.voltage, edge.from_, edge.to,
                                     self.place_names[edge.source], self.place_names[edge.destination]])


def main():
    g = SmartGrid(count_destinations_from_file())

    read_places_from_file(g)
    read_edges_from_file(g)
    g.store_graph_to_file("initial_grid.txt")
    read_grid_optimization_from_file(g)
    g.print_grid()
    g.store_graph_to_file("Efficient_grid.txt")
    check_destination_from_grid("Krishnapuram")



def count_destinations_from_file():
    destinations = set()

    with open("edges.txt", 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            destinations.add(row[1])

    return len(destinations) + 1


def read_edges_from_file(g):
    with open("edges.txt", 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            src = row[0]
            des = row[1]
            dist = int(row[2])
            volt = float(row[3])
            from_ = row[4]
            to = row[5]
            if g.contains_value(src) and g.contains_value(des):
                a = g.placeMap[src]
                b = g.placeMap[des]
                g.add_edge(a, b, dist, volt, from_, to)


def read_grid_optimization_from_file(g):
    with open("Additionaledges.txt", 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            src = row[0]
            des = row[1]
            dist = int(row[2])
            if g.contains_value(src) and g.contains_value(des):
                a = g.placeMap[src]
                b = g.placeMap[des]
                g.grid_optimization(a, b, dist)


def read_places_from_file(g):
    num = 1
    g.place_names[0] = "grid"
    with open("edges.txt", 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            des = row[1]
            if des not in g.placeMap:
                g.add_place(des, num)
                g.place_names[num] = des
                num += 1

def check_destination_from_grid(d):
    try:
        with open("new_dataset.txt", "r") as file:
            found = False

            for line in file:
                parts = line.strip().split(",")

                if len(parts) >= 2:
                    des = parts[1]
                    src = parts[0]
                    if des == d:
                        found = True  # Update the flag when the destination is found
                        break

            if found:
                print("destination found:", d)
            else:
                print("destination not found")

    except FileNotFoundError as e:
        print(e)




if __name__ == "__main__":
    main()

File: src/main/java/smartgrid/service/SmartGrid.java
package main.java.smartgrid.service;

import main.java.smartgrid.model.Container;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.*;

public class SmartGrid {
    List<List<Container>> Adjlist;
    private int places;
    int[] path;
    double[] voltpath;
    int temp_distance = 0;
    public Map<String, Integer> placeMap = new HashMap<>();
    Map<Integer, Double> voltdict = new HashMap<>();
    public String[] place_names;
    double[] heuristic;

    public SmartGrid(int vertices) {
        this.places = vertices;
        this.Adjlist = new ArrayList<>(vertices);
        this.path = new int[vertices + 1];
        this.voltpath = new double[vertices + 1];
        this.place_names = new String[vertices];
        this.heuristic = new double[vertices];

        // Initialize voltage dictionary
        voltdict.put(30, 33.0);
        voltdict.put(50, 48.0);
        voltdict.put(60, 50.0);
        voltdict.put(80, 60.0);
        voltdict.put(90, 66.0);
        voltdict.put(20, 0.44);
        voltdict.put(70, 49.0);
        voltdict.put(40, 40.0);
        voltdict.put(85, 63.0);
        voltdict.put(120, 345.0);
        voltdict.put(100, 345.0);
        voltdict.put(200, 400.0);
        voltdict.put(300, 450.0);
        voltdict.put(400, 490.0);
        voltdict.put(190, 380.0);
        voltdict.put(160, 360.0);

        placeMap.put("grid", 0);

        for (int i = 0; i < vertices; i++) {
            this.Adjlist.add(new ArrayList<>());
        }

        for (int i = 0; i < vertices; i++) {
            heuristic[i] = 0; // Will be updated externally
        }
    }

    public void addEdge(int src, int des, int dist, double volt, char from, char to) {
        Container c = new Container(src, des, dist, volt, from, to);
        Adjlist.get(src).add(c);
    }

    public void removeEdge(int src, int des) {
        Iterator<Container> iterator = Adjlist.get(src).iterator();
        while (iterator.hasNext()) {
            Container neigh = iterator.next();
            if (neigh.destination == des) {
                iterator.remove();
                System.out.println("Edge removed: " + src + " --- " + des);
                break;
            }
        }
    }

    public int First_findpath(int src, int des, int p_src) {
        for (Container neighbor : Adjlist.get(src)) {
            temp_distance += neighbor.distance;
            path[p_src] = neighbor.destination;
            voltpath[p_src] = neighbor.voltage;
            if (neighbor.destination == des) {
                return temp_distance;
            }
            int dist = First_findpath(neighbor.destination, des, p_src + 1);
            if (dist == 0) {
                temp_distance -= neighbor.distance;
                path[p_src] = 0;
                voltpath[p_src] = 0;
            } else {
                return dist;
            }
        }
        return 0;
    }

    public int Second_findpath(int src, int des) {
        for (Container neighbor : Adjlist.get(src)) {
            temp_distance += neighbor.distance;

            if (neighbor.destination == des) {
                return temp_distance;
            }
            int dist = Second_findpath(neighbor.destination, des);
            if (dist == 0) {
                temp_distance -= neighbor.distance;
            } else {
                return dist;
            }
        }
        return 0;
    }

    public void updateHeuristics(int goal) {
        for (int i = 0; i < places; i++) {
            heuristic[i] = estimateDistance(i, goal);
        }
    }

    private int estimateDistance(int from, int to) {
        boolean[] visited = new boolean[places];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] { from, 0 });
        visited[from] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int node = current[0];
            int dist = current[1];

            if (node == to)
                return dist;

            for (Container edge : Adjlist.get(node)) {
                if (!visited[edge.destination]) {
                    visited[edge.destination] = true;
                    queue.add(new int[] { edge.destination, dist + edge.distance });
                }
            }
        }
        return Integer.MAX_VALUE;
    }

    public void GridOptimization(int s, int d, int dist) {
        Arrays.fill(path, 0);
        Arrays.fill(voltpath, 0.0);

        double getvolt = voltdict.getOrDefault(dist, 0.0);

        updateHeuristics(d); // update heuristic values for A*

        int min_distance = AStarPath(0, d);
        temp_distance = 0;
        int org_dis = Second_findpath(0, s);
        int sum = org_dis + dist;

        if (sum > min_distance) {
            System.out.println("Min-distance: 0 -- " + d + " is " + min_distance);
        } else {
            System.out.println("Min-distance: 0 -- " + d + " is " + sum);
            int j = places - 1;
            while (j > 0) {
                if (path[j] == 0) {
                    j--;
                } else {
                    removeEdge(path[j - 1], path[j]);
                    addEdge(s, d, dist, getvolt, 'c', 'c');
                    System.out.println("Edge added: " + s + " --- " + d);
                    j = 0;
                }
            }
        }
    }

    public int AStarPath(int start, int goal) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        boolean[] visited = new boolean[places];
        pq.add(new int[] { start, 0 });

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int node = current[0];
            int gCost = current[1];

            if (node == goal)
                return gCost;

            if (visited[node])
                continue;
            visited[node] = true;

            for (Container edge : Adjlist.get(node)) {
                if (!visited[edge.destination]) {
                    int fCost = gCost + edge.distance + (int) heuristic[edge.destination];
                    pq.add(new int[] { edge.destination, fCost });
                }
            }
        }
        return Integer.MAX_VALUE;
    }

    public void printGrid() {
        System.out.println("Source  Destination  Distance  Voltage");
        for (int i = 0; i < places; i++) {
            for (Container neigh : Adjlist.get(i)) {
                System.out.println("(" + neigh.source + ", " + neigh.destination + ") --> " +
                        neigh.distance + " <--> " + neigh.voltage);
            }
        }
        for (String key : placeMap.keySet()) {
            System.out.println("Place: " + key + ", Index: " + placeMap.get(key));
        }
    }

    public boolean containsValue(String key) {
        return placeMap.containsKey(key);
    }

    public void addplace(String name, int num) {
        placeMap.put(name, num);
    }

    public void storeGraphToFile(String filename) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (int i = 0; i < Adjlist.size(); i++) {
                List<Container> edges = Adjlist.get(i);
                for (Container edge : edges) {
                    writer.write(edge.source + "," + edge.destination + "," + edge.distance + "," +
                            edge.voltage + "," + edge.from + "," + edge.to + "," + place_names[edge.source] + "," +
                            place_names[edge.destination]);
                    writer.newLine();
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
